dev/com.ibm.ws.concurrent.mp.cdi/src/com/ibm/ws/concurrent/mp/cdi/context/WeldContextProvider.java

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.concurrent.mp.cdi.context;
+
+import java.util.Map;
+
+import org.eclipse.microprofile.concurrent.ThreadContext;
+import org.eclipse.microprofile.concurrent.spi.ThreadContextController;
+import org.eclipse.microprofile.concurrent.spi.ThreadContextProvider;
+import org.eclipse.microprofile.concurrent.spi.ThreadContextSnapshot;
+import org.jboss.weld.manager.api.WeldManager;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import com.ibm.websphere.ras.annotation.Trivial;
+import com.ibm.ws.cdi.CDIService;
+
+@Component(immediate = true, name = "com.ibm.ws.cdi.context.provider")
+public class WeldContextProvider implements ThreadContextProvider {
+
+    private static final NoopSnapshot NO_OP_SNAPSHOT = new NoopSnapshot();
+
+    @Reference
+    CDIService cdiService;
+
+    @Override
+    public ThreadContextSnapshot currentContext(Map<String, String> props) {
+        if (!cdiService.isCurrentModuleCDIEnabled())
+            return NO_OP_SNAPSHOT;
+        return new WeldContextSnapshot(true, (WeldManager) cdiService.getCurrentModuleBeanManager());
+    }
+
+    @Override
+    public ThreadContextSnapshot clearedContext(Map<String, String> props) {
+        if (!cdiService.isCurrentModuleCDIEnabled())
+            return NO_OP_SNAPSHOT;
+        return new WeldContextSnapshot(false, (WeldManager) cdiService.getCurrentModuleBeanManager());
+    }
+
+    @Override
+    @Trivial
+    public String getThreadContextType() {
+        return ThreadContext.CDI;
+    }
+
+    @Trivial
+    private static class NoopSnapshot implements ThreadContextSnapshot, ThreadContextController {
```

Comment:
I don't understand the No-op snapshot.  I can understand that CDI might not be available from the code that creates the task/action.  However, wouldn't it be possible that different code which runs the task/action later might have a valid CDI context, which a No-op snapshot would erroneously leave in place.

Reply:
I will have to investigate this. The only time I could imagine this happening is if a task is submitted in one Bean Deployment Archive (BDA) which is not CDI enabled, and then the task runs in a different BDA which is CDI enabled.
If this is possible, then we would also need to consider the reverse scenario, where we go from CDI-enabled BDA to a non-CDI BDA.

Reply:
Could this happen when a servlet invokes an EJB where one enables CDI but the other does not?
In any case, that work can be done under a separate delivery, as can addressing other review comments, which are all minor.  There is already plenty under this pull, and no reason to hold up progress given that the code isn't GA.

Reply:
What do you state in the spec on the scenario where one module is cdi enabled while the other does not? How does the propagation work?